src/wasm-linker.cpp

```diff
@@ -49,13 +49,38 @@ void Linker::placeStackPointer(size_t stackAllocation) {
 }
 
 void Linker::layout() {
+  // Convert calls to undefined functions to call_imports
+  for (const auto& f : out.undefinedFunctions) {
+    Name target = f.first;
+    // Create an import for the target if necessary.
+    if (!out.wasm.checkImport(target)) {
+      auto import = out.wasm.allocator.alloc<Import>();
+      import->name = import->base = target;
+      import->module = ENV;
+      import->type = ensureFunctionType(getSig(*f.second.begin()), &out.wasm,
+                                        out.wasm.allocator);
+      out.wasm.addImport(import);
+    }
+    // Change each call. The target is the same since it's still the name.
+    // Delete and re-allocate the Expression as CallImport to avoid undefined
+    // behavior.
+    for (auto* call : f.second) {
+      Call callCopy = std::move(*call);
+      call->~Call();
```

Comment:
Destructors are unneeded - we allocate in areas, which are destroyed all at once. As a result, arena-allocated classes can't have destructors, as they wouldn't be called. (This causes some leaks currently, which I have a patch in progress for.)

In other words, when you don't need something in an arena, just forget about it.